keps/sig-apps/20200302-surge-on-daemonsets.md

```diff
@@ -0,0 +1,148 @@
+
+As an optimization for users we can consider rejecting via validation changes that would prevent scheduling of surging daemonsets:
+
+* HostPort use when MaxSurge != 0
+
+However in general we do not perform this validation in other contexts, and its novelty probably does not justify the extra complexity.
+
+### Workload Implications
+
+There are three main workload types that seek to minimize disruption:
+
+1. Infrastructure that should be quickly replaced during update (CNI plugins, CSI plugins).
+2. Infrastructure that wishes to hand off a node resource during an upgrade (socket, namespace, process)
+3. Infrastructure that must remain 100% available to support workloads (networking components, proxies).
+
+In general, all of these benefit from minimizing the time between old pod shutting down and new pod starting up. MaxSurge allows components to arbitrarily approach zero disruption by careful tuning of their launch scripts and access to shared resources, such as sockets or shared disk.
+
+Infrastructure invoked by Kubernetes components (CRI, CNI, CSI) can usually fall within the first category and may require some coordination from the invoking process to minimize downtime. For instance, the Kubelet may retry certain types of CSI errors transparently to mitigate brief disruption to a CSI plugin. Or the container runtime may retry certain CNI errors if the plugin is not available.
+
+The second category of workload requires some coordination between the old and new container - for instance, reusing a host volume and checking for file locking on shared resources, or using the SO_REUSEPORT option to start listening on an interface and share old and new traffic. In general the workload author is assumed to understand how to minimize disruption and Kubernetes is only giving them an overlapping window of execution before beginning the termination of the old process.
+
+The last category category is the most difficult to achieve and generally combines categories 1 and 2 along with careful tuning. Networking plugins that provide pod network capability may have one or more daemon processes that are desirable to deliver containerized, but any disruption to those critical pods may impact other workloads. In most cases, the capability to overlap execution provided by the MaxSurge is sufficient to allow those components to adapt to zero-downtime updates.
+
+### Risks and Mitigations
```

Comment:
if this is proposing allowing additional values for MaxUnavailable, describe the phased rollout that ensures we never allow DaemonSet data to persist which an active API server in the cluster would consider invalid

Reply:
Covered, let me know what you think.